File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionResizeTo.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionResizeTo extends SKAction {
    private SKSize startSize;
    private SKSize deltaSize;

    public SKActionResizeTo(SKSize deltaSize) {
        this.deltaSize = deltaSize;
    }

    @Override
    void computeStart() {
        if (getParent() instanceof SKSpriteNode) {
            startSize = ((SKSpriteNode)getParent()).getSize();
        }
    }

    @Override
    void computeAction(long elapsed) {
        if (getParent() instanceof SKSpriteNode) {
            SKSpriteNode sNode = ((SKSpriteNode) getParent());
            float newWidth = sNode.getSize().width;
            float newHeight = sNode.getSize().height;
            switch (timingMode) {
                case Linear:
                    newWidth = SKEaseCalculations.linear(elapsed, startSize.width, deltaSize.width - startSize.width, duration);
                    newHeight = SKEaseCalculations.linear(elapsed, startSize.height, deltaSize.height - startSize.height, duration);
                    break;
                case EaseIn:
                    newWidth = SKEaseCalculations.easeIn(elapsed, startSize.width, deltaSize.width - startSize.width, duration);
                    newHeight = SKEaseCalculations.easeIn(elapsed, startSize.height, deltaSize.height - startSize.height, duration);
                    break;
                case EaseOut:
                    newWidth = SKEaseCalculations.easeOut(elapsed, startSize.width, deltaSize.width - startSize.width, duration);
                    newHeight = SKEaseCalculations.easeOut(elapsed, startSize.height, deltaSize.height - startSize.height, duration);
                    break;
                case EaseInEaseOut:
                    newWidth = SKEaseCalculations.easeInOut(elapsed, startSize.width, deltaSize.width - startSize.width, duration);
                    newHeight = SKEaseCalculations.easeInOut(elapsed, startSize.height, deltaSize.height - startSize.height, duration);
                    break;
            }
            sNode.setSize(new SKSize(newWidth, newHeight));
        }
    }

    @Override
    void computeFinish() {
        if (getParent() instanceof SKSpriteNode) {
            ((SKSpriteNode)getParent()).setSize(new SKSize(deltaSize.width, deltaSize.height));
        }
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionResizeTo.kt
package br.com.insanitech.spritekit

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionResizeTo(deltaSize: SKSize) : SKAction() {
    private val startSize = SKSize()
    private val deltaSize = SKSize()

    init {
        this.deltaSize.assignByValue(deltaSize)
    }

    internal override fun computeStart() {
        (parent as? SKSpriteNode ?: return).let { parent ->
            startSize.assignByValue(parent.size)
        }
    }

    internal override fun computeAction(elapsed: Long) {
        if (parent is SKSpriteNode) {
            val sNode = parent as SKSpriteNode
            var newWidth = sNode.size.width
            var newHeight = sNode.size.height
            when (timingMode) {
                SKActionTimingMode.Linear -> {
                    newWidth = SKEaseCalculations.linear(elapsed.toFloat(), startSize.width, deltaSize.width - startSize.width, duration.toFloat())
                    newHeight = SKEaseCalculations.linear(elapsed.toFloat(), startSize.height, deltaSize.height - startSize.height, duration.toFloat())
                }
                SKActionTimingMode.EaseIn -> {
                    newWidth = SKEaseCalculations.easeIn(elapsed.toFloat(), startSize.width, deltaSize.width - startSize.width, duration.toFloat())
                    newHeight = SKEaseCalculations.easeIn(elapsed.toFloat(), startSize.height, deltaSize.height - startSize.height, duration.toFloat())
                }
                SKActionTimingMode.EaseOut -> {
                    newWidth = SKEaseCalculations.easeOut(elapsed.toFloat(), startSize.width, deltaSize.width - startSize.width, duration.toFloat())
                    newHeight = SKEaseCalculations.easeOut(elapsed.toFloat(), startSize.height, deltaSize.height - startSize.height, duration.toFloat())
                }
                SKActionTimingMode.EaseInEaseOut -> {
                    newWidth = SKEaseCalculations.easeInOut(elapsed.toFloat(), startSize.width, deltaSize.width - startSize.width, duration.toFloat())
                    newHeight = SKEaseCalculations.easeInOut(elapsed.toFloat(), startSize.height, deltaSize.height - startSize.height, duration.toFloat())
                }
            }
            sNode.setSize(newWidth, newHeight)
        }
    }

    internal override fun computeFinish() {
        if (parent is SKSpriteNode) {
            (parent as SKSpriteNode).setSize(deltaSize.width, deltaSize.height)
        }
    }

    internal override fun willHandleFinish(): Boolean {
        return false
    }
}
